### DISCUSSION TITLE
Language Idea: Queryable Iterators

### DISCUSSION THREAD

**Discussion by AdamSobieski** [dotnet/csharplang#7638] [Language Ideas]
## Introduction

Hello. I would like to share and to flesh out some ideas pertaining to an interesting C# language idea: queryable iterators.

## C# Iterators

Quoting from [this article](https://csharpindepth.com/Articles/IteratorBlockImplementation), we can observe that an iterator function resembling:

```cs
class Test
{
    static IEnumerable<int> GetCounter()
    {
        for (int count = 0; count < 10; count++)
        {
            yield return count;
        }
    }
}
```

can be compiled into something like (metadata attributes stripped for brevity):

```cs
internal class Test
{
    private static IEnumerable<int> GetCounter()
    {
        return new <GetCounter>d__0(-2);
    }

    private sealed class <GetCounter>d__0 : IEnumerable<int>, IEnumerable, IEnumerator<int>, IEnumerator, IDisposable
    {
        // Fields
        private int <>1__state;
        private int <>2__current;
        private int <>l__initialThreadId;
        public int <count>5__1;

        public <GetCounter>d__0(int <>1__state)
        {
            this.<>1__state = <>1__state;
            this.<>l__initialThreadId = Thread.CurrentThread.ManagedThreadId;
        }

        private bool MoveNext()
        {
            switch (this.<>1__state)
            {
                case 0:
                    this.<>1__state = -1;
                    this.<count>5__1 = 0;
                    while (this.<count>5__1 < 10)
                    {
                        this.<>2__current = this.<count>5__1;
                        this.<>1__state = 1;
                        return true;
                    Label_0046:
                        this.<>1__state = -1;
                        this.<count>5__1++;
                    }
                    break;

                case 1:
                    goto Label_0046;
            }
            return false;
        }

        IEnumerator<int> IEnumerable<int>.GetEnumerator()
        {
            if ((Thread.CurrentThread.ManagedThreadId == this.<>l__initialThreadId) && (this.<>1__state == -2))
            {
                this.<>1__state = 0;
                return this;
            }
            return new Test.<GetCounter>d__0(0);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable<Int32>) this).GetEnumerator();
        }

        void IEnumerator.Reset()
        {
            throw new NotSupportedException();
        }

        void IDisposable.Dispose()
        {
        }

        int IEnumerator<int>.Current
        {
            get
            {
                return this.<>2__current;
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return this.<>2__current;
            }
        }
    }
}
```

## Queryable Iterators

What if developers could express, in hopes of enabling runtime optimizations, something like:

```cs
class Test
{
    static IQueryable<int> GetCounter()
    {
        for (int count = 0; count < 10; count++)
        {
            yield return count;
        }
    }
}
```

In this case, the resultant compiled code could be such so that the private sealed class `<GetCounter>d__0` would also implement `IQueryable<int>` and, thus, `IQueryable`.

This would entail that the class `<GetCounter>d__0` would have properties [`ElementType`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.iqueryable.elementtype?view=net-7.0#system-linq-iqueryable-elementtype), [`Expression`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.iqueryable.expression?view=net-7.0#system-linq-iqueryable-expression), and [`Provider`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.iqueryable.provider?view=net-7.0#system-linq-iqueryable-provider).

The matter of implementing `ElementType` would be simple. For the example, it would return `typeof(int)`.

For the `Provider` property, a resulting private sealed class `<GetCounter>d__0` could, if its private `IQueryProvider` member, e.g., `<>n__provider`, were `null`, activate one using a runtime-default provider type, caching the value. In theory, .NET could provide a default query provider for queryable iterators. As described below, developers could provide custom ones or use third-party ones.

The remaining important matter for consideration, then, would be the nature of the expression tree returned by the `Expression` property. It could represent: (1) the iterator method before the compiler-time transformation, (2) an `InvocationExpression` which invokes a useful `LambdaExpression`, (3) a `MethodCallExpression` expression, (4) another, perhaps more intricate, technique.

## Customizable Query Providers

Developers using queryable iterators might want to provide their own `IQueryProvider` implementations or to make use of third-party ones. This could be achieved with metadata:

```cs
class Test
{
    [QueryProvider(typeof(CustomQueryProvider))]
    static IQueryable<int> GetCounter()
    {
        for (int count = 0; count < 10; count++)
        {
            yield return count;
        }
    }
}
```

That metadata attribute, or another identically initialized attribute, could be placed on the resulting private sealed class `<GetCounter>d__0`. The `<GetCounter>d__0` class could, for its `Provider` property, if a private backing field, `<>n__provider`, were `null`, activate an instance of that type obtained from its `QueryProviderAttribute` metadata attribute and assign it to that backing field member, caching the value.

## Use Case: C# Logic Programming

Optimizations needn't entirely focus on the streaming outputs from the iteration (see: LINQ). Developers could also perform operations and processing upon input arguments provided to iterator functions. This is the case with some approaches to logic programming in C#, e.g., [YieldProlog](https://yieldprolog.sourceforge.net/) (YP).

```cs
IEnumerable<bool> uncle(object Person, object Uncle)
{
    Variable Parent = new Variable();
    foreach (bool l1 in parent(Person, Parent))
    {
        foreach (bool l2 in brother(Parent, Uncle))
            yield return false;
    }
}
```

While perhaps beyond the scope of this discussion about queryable iterators as a language idea, the YP C# logic programs `unify` variables as iterators are iterated, searching for solutions. Outputs can be provided in the form of those instantaneous values on the variables passed to the iterator functions as arguments. Each time that `MoveNext()` is called on the resultant enumerator, the variables could have different values.

This is a different scenario than that of streaming output values (see: LINQ) but is also one that a third-party library provider might find interesting to provide optimizations for.

### Enumerables

Optimizations are possible with respect to queryable iterators and LINQ operators. Let us consider an optimization-related extension method, `Optimize<T>()`, resembling the following:

```cs
public static IQueryable<T> Optimize<T>(this IQueryable<T> source, IOptimizationConfiguration config)
{
    if (source.Provider is IOptimizingQueryProvider source_optimizer)
    {
        return source_optimizer.Optimize(source.Expression, config);
    }
    else
    {
        var default_optimizer = DefaultOptimizer.Create(config);
        return default_optimizer.Optimize(source);
    }
}
```

Let us also consider the following two queryable iterator functions:

```cs
public IQueryable<int> Function1()
{
    for(int i = 0; i < 10000; ++i)
        yield return i;
}
```

and

```cs
public IQueryable<int> Function2()
{
    for(int i = 10000; i < 20000; ++i)
        yield return i;
}
```

With the following LINQ:

```cs
var result = Function1().Concat(Function2()).Optimize(config);
```

a resultant queryable iterator function, `Function3`, might resemble:

```cs
public IQueryable<int> Function3()
{
    for(int i = 0; i < 10000; ++i)
        yield return i;

    for(int i = 10000; i < 20000; ++i)
        yield return i;
}
```

or, if the optimizer detected this edge case, `Function3` might resemble:

```cs
public IQueryable<int> Function3()
{
    for(int i = 0; i < 20000; ++i)
        yield return i;
}
```

Next, utilizing the `Where` operator:

```cs
var result = Function1().Where(n => n % 2 == 0).Optimize(config);
```

a resultant queryable iterator function, `Function4`, might resemble:

```cs
public IQueryable<int> Function4()
{
    for(int i = 0; i < 10000; ++i)
        if(i % 2 == 0)
            yield return i;
}
```

Similarly, with a combined `Where` and `Select` operator:

```cs
var result = Function1().Where(n => n % 2 == 0).Select(n => n * n).Optimize(config);
```

a resultant queryable iterator function, `Function5`, might resemble:

```cs
public IQueryable<int> Function5()
{
    for(int i = 0; i < 10000; ++i)
        if(i % 2 == 0)
            yield return i * i;
}
```

For a more complex LINQ example, let us add another function:

```cs
public IQueryable<int> Function6(int x)
{
    for(int i = x - 10; i < x + 10; ++i)
        yield return i;
}
```

And a LINQ expression:

```cs
var result = Function1().SelectMany(n => Function6(n), (x, y) => {x, y}).Where(p => p.x > p.y).Select(p => p.x * p.y)
    .Optimize(config);
```

a resultant queryable iterator function, `Function7`, might resemble:

```cs
public IQueryable<int> Function7()
{
    for(int i = 0; i < 10000; ++i)
    {
        for(int j = i - 10; j < i + 10; ++j)
        {
            if(i > j)
            {
                return i * j;
            }
        }
    }
}
```

With `LambdaExpression`s available for queryable iterator functions, runtime compilers could produce resultant queryable iterator functions which optimize for, in some cases inlining, those lambda expressions provided for subsequently utilized LINQ operators including, but not limited to, `Concat`, `Where`, `Select`, and `SelectMany`.

### Non-enumerables

Beyond those LINQ operators like `Concat`, `Where`, `Select`, and `SelectMany`, indicated above, one might wonder about those LINQ operators like `All`, `Any`, and `Count`?

In these regards, we might consider `QueryableValue<T>` and `QueryableCachingValue<T>`, classes with which to express "`IQueryable` for non-enumerables". These would provide non-caching and caching access to object values while also providing `ElementType`, `Provider`, and `Expression` properties.

As envisioned, `QueryableValue<T>` would consult its backing method each time its value was requested and `QueryableCachingValue<T>` would cache its value after it was first accessed.

```cs
public static partial class QueryableValue
{
    public static QueryableValue<bool> All<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }
    public static QueryableValue<bool> Any<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }

    public static QueryableValue<int> Count<T>(this IQueryable<T> source) { ... }
    public static QueryableValue<int> Count<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }

    ...
}
```

```cs
public static partial class QueryableCachingValue
{
    public static QueryableCachingValue<bool> All<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }
    public static QueryableCachingValue<bool> Any<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }

    public static QueryableCachingValue<int> Count<T>(this IQueryable<T> source) { ... }
    public static QueryableCachingValue<int> Count<T>(this IQueryable<T> source, Expression<Func<T, bool>> predicate) { ... }

    ...
}
```

With classes resembling those, "`IQueryable` for non-enumerables", the `Optimize()` pattern could be expanded to include more of the LINQ operators.

That is, something like:

```cs
var result = Function1().Where(n => n % 2 == 0).Count().Optimize(config);
```

could result in a function resembling `Function8`, below, being invoked whenever the `Value` is desired:

```cs
public QueryableValue<int> Function8()
{
    int count = 0;

    for(int i = 0; i < 10000; ++i)
        if(i % 2 == 0)
            ++count;

    return count;
}
```

or only when first desired:

```cs
public QueryableCachingValue<int> Function8()
{
    int count = 0;

    for(int i = 0; i < 10000; ++i)
        if(i % 2 == 0)
            ++count;

    return count;
}
```

## Potential Complexities

Some complexities may arise from the LINQ AST not supporting expressiveness for all of the newest C# language features.

The simplest approach to having an inspectable expression tree for a queryable iterator is to use an `InvocationExpression` upon a `LambdaExpression` which represents the program logic of the pre-compiler-transformed queryable iterator. However, as it is an iterator, it utilizes `yield return`, and there is not a `YieldReturnExpression` in the LINQ AST. Accordingly, other approaches can be explored.

## Other Uses and Related Scenarios

Beyond developers hoping to optimize queryable iterator functions at runtime with LINQ, queryable iterators would be a convenient way to easily create useful `IQueryable<T>` instances.

Other, related scenarios include considering return types of `IObservable<T>` and, in particular, `IQbservable<T>` as well as mappings between `IEnumerable<T>` and `IObservable<T>` and `IQueryable<T>` and `IQbservable<T>`. These mappings would be desired to work and to be interoperable with queryable iterators.

```cs
var result = Function1().ToQbservable();
```

## Conclusion

Thank you.

**Comment by CyrusNajmabadi** [comment]
> What if developers could express, in hopes of enabling runtime optimizations, 

What optimizations would this enable. 

Generally speaking, we don't do language features based on speculation that there will be performance. Rather, we want to see that performance is lacking, and has no way to obtain, and then we design with a drive to solve that performance issue. 

This allows us to actually validate the language/solution, and not end up with something that doesn't actually work in practice.

**Reply by AdamSobieski** [reply]
That makes sense. Rather than a single iterator or a few independent iterators, I'm considering scenarios with multiple, nested queryable iterators. I'm thinking that a query provider should be able to process iterators' expression trees to produce output iterators with inlined program logic.

Next week, I can use a disassembler and decompiler to analyze some iterator functions, iterators which call other iterators, nested iterators, and explore what the compiler produces for these, e.g., comparing non-inlined and inlined variations of the same functionalities. After that, I can reply back here with that data and, hopefully, a few more useful source-code examples.

**Reply by GabeSchaffer** [reply]
I have a database with an in-memory cache. If the data I need is in the cache, I can run LINQ queries against it using `Func`s; otherwise I have to send the query to the DB, in which case I need `Expressions`s. If I pass `Func`s to my query methods, I can't pass the queries through to the DB if the data's not in the cache; if I pass in `Expression`s, they have to be created and compiled for each query, negating some of the benefit of caching.

In order to get the best performance out of my cache, I end up having to write my expressions as both `Func`s and `Expression`s, even though the code is identical.

If I had a way to ask a `Func` "what is your representation as an `Expression`?", I could write my expressions once and have a query engine that executes in-memory queries as fast as possible while still being able to automatically send queries it can't fulfill to the DB.

**Reply by AdamSobieski** [reply]
While creating a source-code example for the possibilities:

> (1) the iterator method before the compiler-time transformation, (2) an `InvocationExpression` which invokes a useful `LambdaExpression`

I found and remembered that there is no `YieldReturnExpression` in LINQ.

That is, were it possible to form the `invoke` expression, as constructed here:

```cs
var y1 = Expression.Yield(Expression.Constant(1));
var y2 = Expression.Yield(Expression.Constant(2));
var y3 = Expression.Yield(Expression.Constant(3));
var b1 = Expression.Block(typeof(IQueryable<int>), y1, y2, y3);
var lambda = Expression.Lambda(b1, "Function1", Array.Empty<ParameterExpression>());
var invoke = Expression.Invoke(lambda);
```

then that `invoke` could have been the value for the `Expression` property on the `IQueryable<int>` returned by:

```cs
public IQueryable<int> Function1()
{
  yield return 1;
  yield return 2;
  yield return 3;
}
```

that is,

```cs
var expr = Function1().Expression;
```

And, introducing a second function:

```cs
public IQueryable<int> Function2()
{
  yield return 4;
  yield return 5;
  yield return 6;
}
```

I was thinking that one of the first, simplest examples of an optimization might have involved:

```cs
var result = Function1().Concat(Function2());
```

which would have, instead of iterating one and then the other, produced, upon being iterated, an optimized iterator which would yield 6 results.

```cs
public IQueryable<int> Function3()
{
  yield return 1;
  yield return 2;
  yield return 3;
  yield return 4;
  yield return 5;
  yield return 6;
}
```

A simple example... yes... but, the obstacle to options (1) and (2), when the invoked lambda is desired to be the pre-transformed method, is the absence of a `YieldReturnExpression`.

Also, onto a different subtopic, it could be the case that more complex optimizations to be described, upcoming, optimizations to be performed by query providers, could require optimization hints in the form of metadata.

```cs
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public IQueryable<int> Function1()
{
  yield return 1;
  yield return 2;
  yield return 3;
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public IQueryable<int> Function2()
{
  yield return 4;
  yield return 5;
  yield return 6;
}
```

Alternatively or additionally, it could be the case that an `Optimize()` extension method would be configurable by developers.

```cs
var result = Function1().Concat(Function2()).Optimize(config);
```

Unless the absence of a `YieldReturnExpression` is entirely prohibitive to solving this, I should have a more detailed reply by next week with more examples of potential optimizations and more source-code examples.

**Reply by HaloFour** [reply]
@GabeSchaffer 

If you pass `Expression<...>` to the methods, you can easily keep them as expressions, and/or compile them into lambdas.  That gets you the best of both worlds.  For a while that's what Entity Framework would do.  It would translate part of the query, up to a manageable projection, into SQL.  It would then translate the remainder of the query into lambdas that it would apply to the results.

**Reply by HaloFour** [reply]
Having the compiler emit expression trees for arbitrary methods would be interesting.  I'm not sure how practical it would be, or in how many applications it would actually provide any real value.  However, the massive problem with that approach is due to versioning issues the expression APIs are effectively frozen to C# 3.0.  No decent way forward has been identified.  It's also been suggested that, if designed today, it would be extremely different.  But short of identifying some very specific and widely applicable use case (not "potential optimizations"), I doubt there is any interest in completely redesigning any of that, or reworking the compiler to emit such general purpose expression trees that no existing provider could understand.

**Reply by colejohnson66** [reply]
@HaloFour What do you mean "versioning issues"? An expression is just an abstract object. What's the reason that new derived classes can't be added in later versions?

**Reply by HaloFour** [reply]
Adding new nodes breaks the contract and existing clients.

See: #158

**Reply by CyrusNajmabadi** [reply]
@GabeSchaffer 

> If I had a way to ask a Func "what is your representation as an Expression?", I could write my expressions once 

That seems like a different request than this one. Could you create a new discussion on that? Fwiw, that sounds like a fine idea. 

That said, you can always call Compile to get the func from your `Expression<tfunc>`, so I'm not sure why that's not sufficient. We can discuss it more in the separate discussion. Thanks!

**Reply by CyrusNajmabadi** [reply]
I looked, and I could not find usages of YieldProlog. Do you have other systems that are more widespread that need this functionality to be viable?

**Reply by AdamSobieski** [reply]
I will, then, focus the forthcoming source-code examples on the use cases with streaming iteration values, e.g., LINQ-like iterators, and not as much, if at all, on the logic-programming / searching for variable tuples use cases, e.g., YP-like iterators.

Beyond the optimizing, or inlining, of simple concatenation scenarios, indicated above, here is another, more useful, example: inlining filters.

That is, with something like:

```cs
public IQueryable<int> Function4()
{
  for(int i = 0; i < 10000; ++i)
    yield return i;
}
```

and some LINQ:

```cs
var result = Function4().Where(n => n % 2 == 0).Optimize(config);
```

we can envision a resultant optimized iterator function, `Function5`, which would inline the `Where` functor:

```cs
public IQueryable<int> Function5()
{
  for(int i = 0; i < 10000; ++i)
    if(i % 2 == 0)
      yield return i;
}
```

**Reply by HaloFour** [reply]
Why would you want to inline filters in that way?  Doing that at runtime via introspection sounds like it would be a _lot_ more expensive than nested iterators.

**Reply by CyrusNajmabadi** [reply]
I'm not opposed to this idea. However, it seems like linq would be the place to start with this. Given that that is almost 20 years old, but hasn't really gone that route, it seems less likely we would do anything for yield iterators (which are the much less common case, by several orders of magnitude).

**Reply by AdamSobieski** [reply]
@HaloFour, as considered, there would be runtime-compile-related overhead and, thereafter, each time that the resultant iterator functions were run, a speedup. Long-running software applications, e.g., servers and services, could be benefitted.

Also, in addition to `IEnumerable<T>` and `IQueryable<T>`, there are `IObservable<T>`, and `IQbservable<T>` and `Async` varieties of all of these. The [Reactive libraries](https://github.com/dotnet/reactive) include methods with which to transform between these pull- and push-based models. Maybe optimization ideas for `IQueryable<T>` could benefit `IQbservable<T>`.

Brainstorming, runtime optimizations needn't be only provided by the `Optimize()` extension method. There could be an interface for `IQueryProvider` instances, `IOptimizingQueryProvider`, with an `Optimize()` method on it.

```cs
public static IQueryable<T> Optimize<T>(this IQueryable<T> source, IOptimizationConfiguration config)
{
  if(source.Provider is IOptimizingQueryProvider o)
  {
    return o.Optimize(source.Expression, config);
  }
  else
  {
    var optimizer = DefaultOptimizer.Create(config);
    return optimizer.Optimize(source);
  }
}
```

> I'm not opposed to this idea. However, it seems like linq would be the place to start with this.

Excellent. Ok, I will focus on LINQ-related scenarios for forthcoming source-code examples.

**Reply by AdamSobieski** [reply]
I added some LINQ-related source-code examples to the proposal post. In the revised proposal, I broach optimizations for enumerable-related scenarios like `Concat`, `Select`, `SelectMany`, and `Where`. I also broach optimization ideas for non-enumerable-related scenarios like `All`, `Any`, and `Count`. What do you think?

**Reply by CyrusNajmabadi** [reply]
> I added some LINQ-related source-code examples to the proposal post. In the revised proposal, I broach optimizations for enumerable-related scenarios like Concat, Select, SelectMany, and Where.

I see the examples, but what I don't see in the entire post is: "why".

Because, *"why"* would we do this? This is asking for a tremendous amount of design and work, without any reason given for why this is actually a good thing. 

Literally zero real world cases have been brought up besides 'YieldProlog' (a completely used library). And even for that library no real world examples were made for how this would actually help. 

As such, this seems like a post made just for intellectual exploration. While that can be fun, and I understand the interest in some to explore that realm, it really isn't driven by the thing we care most about: is this valuable to the ecosystem. 

Given a week of discussion, and no real world practical value, it definitely feels like yours is more to please the mind versus provide value.

**Reply by AdamSobieski** [reply]
Yes, I see that adding these language features would entail a significant amount of design and development. I will put some thought into more compelling elevator pitches or sales pitches for queryable iterators, beyond enabling runtime compiler optimizations. I sense that these ideas aren't as popular here as I thought that they would be. Ah well, back to the drawing board.

**Reply by CyrusNajmabadi** [reply]
> I sense that these ideas aren't as popular here 

Agreed. Speculative value, and no real customer identified, will always take a backseat to real gains, with tons of customers indicating support and presenting real world usecases where the feature would be beneficial. 

If you find anything in the real world that would benefit, def let us know. Thanks!
